codeflash/benchmarking/codeflash_trace.py

```diff
@@ -0,0 +1,78 @@
+import functools
+import os
+import dill as pickle
+import time
+from typing import Callable
+
+
+
+
+class CodeflashTrace:
+    """A class that provides both a decorator for tracing function calls
+    and a context manager for managing the tracing data lifecycle.
+    """
+
+    def __init__(self) -> None:
+        self.function_calls_data = []
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        # Cleanup is optional here
+        pass
+
+    def __call__(self, func: Callable) -> Callable:
+        """Use as a decorator to trace function execution.
+
+        Args:
+            func: The function to be decorated
+
+        Returns:
+            The wrapped function
+
+        """
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            # Measure execution time
+            start_time = time.thread_time_ns()
+            result = func(*args, **kwargs)
+            end_time = time.thread_time_ns()
+            # Calculate execution time
+            execution_time = end_time - start_time
+
+            # Measure overhead
+            overhead_start_time = time.thread_time_ns()
+
+            try:
+                # Check if currently in pytest benchmark fixture
+                if os.environ.get("CODEFLASH_BENCHMARKING", "False") == "False":
+                    return result
+
+                # Pickle the arguments
+                pickled_args = pickle.dumps(args, protocol=pickle.HIGHEST_PROTOCOL)
```

Comment:
do the pickling logic like in tracer . try pickle first then dill. also set recursion limits etc.